nvdaHelper/vbufBase/backend.cpp

```diff
@@ -138,61 +137,69 @@ void VBufBackend_t::renderThread_terminate() {
-	if(node->updateAncestor) node=node->updateAncestor;
+	auto lock=this->lock.scopedAcquire();
 	if(!isNodeInBuffer(node)) {
 		LOG_DEBUGWARNING(L"Node at "<<node<<L" not in buffer at "<<this);
 		return false;
 	}
-	LOG_DEBUG(L"Invalidating node "<<node->getDebugInfo());
-	this->lock.acquire();
-	bool needsInsert=true;
-	for(VBufStorage_controlFieldNodeList_t::iterator i=invalidSubtreeList.begin();i!=invalidSubtreeList.end();) {
-		VBufStorage_fieldNode_t* existingNode=*i;
-		if(node==existingNode) { //Node already invalidated
-			LOG_DEBUG(L"Node already invalidated");
-			needsInsert=false;
-			break;
-		} else if(isDescendantNode(existingNode,node)) { //An ancestor is already invalidated
-			LOG_DEBUG(L"An ancestor is already invalidated, returning");
-			needsInsert=false;
-			break;
-		} else if(isDescendantNode(node,existingNode)) { //A descedndant has been invalidated
-			LOG_DEBUG(L"removing a descendant node from the invalid nodes");
-			//Remove the old descendant
-			invalidSubtreeList.erase(i++);
-			//If no other descendants have been removed yet, insert the real node here
-			//At the place of the first found descendant
-			if(needsInsert) {
-				invalidSubtreeList.insert(i,node);
-				needsInsert=false;
-			}
-		} else { //Unrelated node
-			++i;
+	// If this node requires its parent to be updated, Invalidate the closest ancestor that does not require its parent to be updated. 
+	while(node->requiresParentUpdate) {
+		node->allowReuseInAncestorUpdate=false;
+		auto parent=node->getParent();
+		if(!parent) break;
+		node=parent;
+		LOG_DEBUG(L"node requiresParentUpdate, therefore trying to invalidate parent at "<<node);
+	}
+	// If this node is already invalidated, do nothing.
+	// If this node is the descendant of a node that is already invalidated,
+	// Ensure that this node and any of its ancestors (up to but not including) the already invalid node, are marked as non-reusable.
+	for(auto i: pendingInvalidSubtreesList) {
+		if(i==node) {
+			return true;
+		} else if(markNodeAsNonreusableIfInAncestor(node,i)) {
+			return true;
 		}
-	}  
-	//If the node still has not been inserted yet, insert it at the end now
-	if(needsInsert) {
-		LOG_DEBUG(L"Adding node to invalid nodes");
-		invalidSubtreeList.insert(invalidSubtreeList.end(),node);
 	}
-	this->lock.release();
+	// If this node is an ancestor of one or more already invalid nodes,
+	// Ensure that the already invalid node and any of its ancestors (up to but not including) this node, are marked as non-reusable. 
+	// Then remove those nodes from the invalidation list.
+	pendingInvalidSubtreesList.remove_if([node](auto i){
+		return markNodeAsNonreusableIfInAncestor(i,node);
+	});
+	// Now mark this node as invalid.
+	pendingInvalidSubtreesList.push_back(node);
+	LOG_DEBUG(L"Invalidated node "<<node->getDebugInfo());
 	this->requestUpdate();
 	return true;
 }
 
 void VBufBackend_t::update() {
+	if(inUpdate) {
+		return;

```

Comment:
Is this to prevent updates requested by win events during a render from running within COM calls? It might be worth commenting the reason for possible re-entry. Also, if this does occur, how does the pending update get re-requested? I guess some other event will fire eventually, but that seems unreliable. Or am I missing something?